api/register/api.go

```diff
@@ -0,0 +1,202 @@
+	regExpBirthLocation = regExpNames
+	// Multiple values with digits and letters separated by a single separator (space, dash)
+	regExpIDDocumentNumber = `^([\w\d]+([ -][\w\d]+)*){1,50}$`
+
+	dateLayout = "02.01.2006"
+)
+
+var (
+	allowedGender       = map[string]bool{"M": true, "F": true}
+	allowedDocumentType = map[string]bool{"Identity card": true, "Passport": true}
+)
+
+// UserFromJSON creates a User using its json representation
+func UserFromJSON(jsonRep string) (User, error) {
+	var user User
+	dec := json.NewDecoder(strings.NewReader(jsonRep))
+	dec.DisallowUnknownFields()
+	err := dec.Decode(&user)
+	return user, err
+}
+
+// UserToJSON returns a json representation of a given User
+func (u *User) UserToJSON() string {
+	var bytes, _ = json.Marshal(u)
+	return string(bytes)
+}
+
+// UpdateUserRepresentation converts a given User to a Keycloak UserRepresentation
+func (u *User) UpdateUserRepresentation(kcUser *kc.UserRepresentation) {
+	var (
+		bFalse     = false
+		attributes = make(map[string][]string)
+	)
+
+	if u.Gender != nil {
+		attributes["gender"] = []string{*u.Gender}
+	}
+	if u.PhoneNumber != nil {
+		attributes["phoneNumber"] = []string{*u.PhoneNumber}
+		attributes["phoneNumberVerified"] = []string{"false"}
+	}
+	if u.BirthDate != nil {
+		attributes["birthDate"] = []string{*u.BirthDate}
+	}
+
+	kcUser.Email = u.EmailAddress
+	kcUser.EmailVerified = &bFalse
+	kcUser.Enabled = &bFalse
+	kcUser.FirstName = u.FirstName
+	kcUser.LastName = u.LastName
+	kcUser.Attributes = &attributes
+}
+
+// Validate checks the validity of the given User
+func (u *User) Validate() error {
+	var err = validateParameterIn(prmUserGender, u.Gender, allowedGender, true)
+	if err != nil {
+		return err
+	}
+	err = validateParameterRegExp(prmUserFirstName, u.FirstName, regExpFirstName, true)
+	if err != nil {
+		return err
+	}
+	err = validateParameterRegExp(prmUserLastName, u.LastName, regExpLastName, true)
+	if err != nil {
+		return err
+	}
+	err = validateParameterRegExp(prmUserEmail, u.EmailAddress, regExpEmail, true)
+	if err != nil {
+		return err
+	}
+	err = validateParameterPhoneNumber(prmUserPhoneNumber, u.PhoneNumber)
+	if err != nil {
+		return err
+	}
+	err = validateParameterDate(prmUserBirthDate, u.BirthDate, false)
+	if err != nil {
+		return err
+	}
+	err = validateParameterRegExp(prmUserBirthLocation, u.BirthLocation, regExpBirthLocation, false)
+	if err != nil {
+		return err
+	}
+	err = validateParameterIn(prmUserIDDocumentType, u.IDDocumentType, allowedDocumentType, false)
+	if err != nil {
+		return err
+	}
+	err = validateParameterRegExp(prmUserIDDocumentNumber, u.IDDocumentNumber, regExpIDDocumentNumber, false)
+	if err != nil {
+		return err
+	}
+	err = validateParameterDate(prmUserIDDocumentExpiration, u.IDDocumentExpiration, false)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func validateParameterIn(prmName string, value *string, allowedValues map[string]bool, mandatory bool) error {
+	if value == nil {
```

Comment:
this function and the following next one are very similar , but I don't see yet if they could be combined in just one function.....

Reply:
It is not the same validation type... not easy to factorize